src/batch-submitter/tx-batch-submitter.ts

```diff
@@ -120,9 +128,18 @@ export class TransactionBatchSubmitter extends BatchSubmitter {
       )
 
       if (!this.disableQueueBatchAppend) {
+        const contractFunction = async (gasPrice): Promise<TransactionReceipt> => {
+          const tx = await this.chainContract.appendQueueBatch(99999999, {gasPrice})
+          return this.signer.provider.waitForTransaction(
+            tx.hash,
+            this.numConfirmations,
+            this.receiptTimeout
+          )
+        }
```

Comment:
I think there isn't an easy way to do this but I really do wish we could de-duplicate these `contractFunction`s -- there's just a lot of boilerplate that we have with this pattern.

So what could make this cleaner is to refactor submitAndLogTransaction to something like:

```
_submitAndLogTx(contractFunction: Function, params: any[], options = {}, logMessage: string) {
  const contractFunction = async (gasPrice): Promise<TransactionReceipt> => {
    options.gasPrice = gasPrice
    params.push(gasPrice)
    const tx = await contractFunction(...params)
    return this.signer.provider.waitForTransaction(....)
  }

  ... the rest of submit and log transaction ...
}
```

And then you can use that function here by just doing:

```
return this._submitAndLogTx(this.chainContract.appendQueueBatch, [99999999], {}, 'My log message!')
```

Reply:
Btw there's definitely better syntax to do what I was trying to express but while not perfectly pretty this would at least remove the duplicate logic^

Reply:
^this sounds good to me! will refactor

Reply:
@karlfloersch thinking through this change right now:

- will `_submitAndLogTx` only submit contract call functions in the future, and are the only parameters possible `params` and `options`? 

my concern is that we over-specialize `_submitAndLogTx` to only contract calls and make it less extensible in the future if this makes other types of txs in the future, but you currently know more abt how this will be used than me so deferring to you here!

another repetition is in `run-batch-submitter`:
https://github.com/ethereum-optimism/batch-submitter/blob/557844c5dd1366d790e8dff299a058fd0c738722/src/exec/run-batch-submitter.ts#L180-L207
where only `getReceiptWithResubmission` is called. right now everything is pretty verbose but consistent, so another tradeoff of specializing `_submitAndLogTx` is that its usage pattern will be pretty different from `getReceiptWithResubmission` (vs how it's basically a wrapper), is that preferred?

Reply:
Hey @karlfloersch, your example above won't exactly work due to  `contractFunction` being declared and being passed as a param. Also the final arg to `params` needs to be an object with overwrites of things on the tx like `gasPrice`, `gas` etc. Right now the `gasPrice` is being passed as an additional argument which will cause the contract call to fail when it asserts the number of args being passed in under the hood.

Reply:
Basically what Karl is suggesting makes sense, its to move the majority of the `contractFunction` into the `submitAndLogTx` function, so that you pass in just the function that is called inside of the `contractFunction` to `submitAndLogTx` and there it will build the `contractFunction` instead of building the `contractFunction` outside and passing it in

Reply:
> your example above won't exactly work due to contractFunction being declared and being passed as a param.

Whoops just updated it -- meant to call the declared function `sendTransactionFunction(...)`

Reply:
@karlfloersch just coming around to reading your commend and it makes sense! the current `getReceiptWithResubmission` behavior is what you described - the function itself times out and errors if the transaction never gets accepted. anything we might do here, like keep the old function under a different name or separating these 2 functions out?

Reply:
@annieke after thinking about it more, I think we can just remove the resubmission logic entirely from `run-batch-submitter` and there just use `wallet.sendTransaction(...)` and `await txRes.wait()`. It's probably fine that we don't have complex resubmission logic for clearing pending transactions on startup

Reply:
So to recap:

1. Ideally we make the calls to `_submitAndLogTx(..)` significantly DRY-er by using something like the function I originally sketched out [here](https://github.com/ethereum-optimism/batch-submitter/pull/22/files#r570743080)
2. We remove the call to `getReceiptWithResubmission(..)` [here](https://github.com/ethereum-optimism/batch-submitter/blob/master/src/exec/run-batch-submitter.ts#L173-L180) and replace it with a simple `await response.wait()`

If you'd like to just ship this PR before DRY-ing the code then I'm very down to say ship it. And then the other option is of course deduplicate the way that we call `_submitAndLogTx(..)` that I originally described and remove `getReceiptWithResubmission(..)`.

Reply:
will do both :)

Reply:
the deduplication change is currently not passing tests, suspect it's a javascript language nuance. the changes are in this draft PR: https://github.com/ethereum-optimism/batch-submitter/pull/26